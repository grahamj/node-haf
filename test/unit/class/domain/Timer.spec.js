const sinon = require('sinon');
const { expect } = require('chai');
const proxyquire = require('proxyquire').noPreserveCache();

describe('Timer class', () => {
  let Timer;
  let constructorStub;
  let callServiceStub;

  beforeEach(() => {
    constructorStub = sinon.stub();
    callServiceStub = sinon.stub();
    class Entity {
      constructor(...args) {
        this.entityId = args[0].entityId;
        constructorStub(...args);
      }
    }
    Entity.prototype.callService = callServiceStub;
    Timer = proxyquire('../../../../src/class/domain/Timer', {
      '../base/Entity.js': Entity,
    });
  });

  describe('constructor()', () => {

    it('Passes config and proper domain to super', () => {
      new Timer({ fancy: true });
      expect(constructorStub.firstCall.args[0]).to.deep.equal({
        fancy: true,
        domain: 'timer',
      });
    });

  });

  describe('start()', () => {

    it('Calls service with proper params, duration specified', async () => {
      const timer = new Timer({ entityId: 'fancy' });
      await timer.start(5);
      expect(callServiceStub.firstCall.args).to.matchSnapshot();
    });

    it('Calls service with proper params, no duration', async () => {
      const timer = new Timer({ entityId: 'fancy' });
      await timer.start();
      expect(callServiceStub.firstCall.args).to.matchSnapshot();
    });

    it('Throws if duration is not a number', async () => {
      const timer = new Timer({ entityId: 'fancy' });
      try {
        await timer.start('abc');
      } catch(err) {
        expect(err.message).to.match(/number/);
        return;
      }
      throw new Error('Should not succeed');
    });

    it('Throws if duration is below 1', async () => {
      const timer = new Timer({ entityId: 'fancy' });
      try {
        await timer.start(0);
      } catch(err) {
        expect(err.message).to.match(/number/);
        return;
      }
      throw new Error('Should not succeed');
    });

  });

  describe('cancel()', () => {

    it('Calls service with proper params', async () => {
      const timer = new Timer({ entityId: 'fancy' });
      await timer.cancel();
      expect(callServiceStub.firstCall.args).to.matchSnapshot();
    });

  });

  describe('finish()', () => {

    it('Calls service with proper params', async () => {
      const timer = new Timer({ entityId: 'fancy' });
      await timer.finish();
      expect(callServiceStub.firstCall.args).to.matchSnapshot();
    });

  });

});
