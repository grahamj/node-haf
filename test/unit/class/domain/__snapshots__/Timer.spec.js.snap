// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer class cancel() Calls service with proper params 1`] = `
Array [
  Object {
    "domain": "timer",
    "service": "cancel",
    "target": Object {
      "entity_id": "fancy",
    },
  },
]
`;

exports[`Timer class finish() Calls service with proper params 1`] = `
Array [
  Object {
    "domain": "timer",
    "service": "finish",
    "target": Object {
      "entity_id": "fancy",
    },
  },
]
`;

exports[`Timer class start() Calls service with proper params, duration specified 1`] = `
Array [
  Object {
    "domain": "timer",
    "service": "start",
    "service_data": Object {
      "duration": 5,
    },
    "target": Object {
      "entity_id": "fancy",
    },
  },
]
`;

exports[`Timer class start() Calls service with proper params, no duration 1`] = `
Array [
  Object {
    "domain": "timer",
    "service": "start",
    "target": Object {
      "entity_id": "fancy",
    },
  },
]
`;
